debian/hpijs.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH HPIJS 1 "2004-05-15"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
hpijs \- HP IJS server for the GhostScript IJS client driver
.SH DESCRIPTION
hpijs should never be called directly.  GhostScript will run it
if the IjsServer parameter is set to IjsServer=hpijs when invoking the
IJS driver.
.br
.SH OPTIONS
.TP
.B -h
Show version
.SH SEE ALSO
For more information, please refer to /usr/share/doc/hpijs, and
to http://www.openprinting.org/.
.TP
/usr/share/doc/hpijs/hpijs_readme.html
.TP
http://hplip.sourceforge.net/
.TP
http://www.openprinting.org/ (search for hpijs driver)
.SH AUTHOR
This manual page was written by Henrique de Moraes Holschuh <hmh@debian.org>,
for the Debian project (but may be used by others).


ChangeLog
2006-01-30 dsuffield
    * Release 0.9.8.22 (public release 0.9.8)
	
2006-01-31 dsuffield
    * Fixed scan port setting for LJ3055 and LJ3060 in xml file.
	
2006-01-30 dsuffield
    * Release 0.9.8.21

2006-01-30 dsuffield
    * Moved HPLIP_02_2006.xml to release directory.

2006-01-30 dsuffield
    * Release 0.9.8.20

2006-01-24 dwelch
    * Release 0.9.8.19

2006-01-24 dwelch
    * Release 0.9.8.18

2006-01-24 dsuffield
    * New APDK label.

2006-01-24 dwelch
    * Release 0.9.8.17

2006-01-23 dwelch
    * Added new io-modes for libusb support for CLJ2500 in model_editor
    * Added new status-types for LJ XML and S:+PML/SNMP in model_editor
    * Fixed the PS32xx/33xx and PS8250 "r_value" issue using new status type 5
    * Cleaned up channel opens/closes in device query for better performance
    
2006-01-19 dsuffield
    * Fixed deviceid zero termination in hplip_api.c.
    * Fixed startup script as per RH request.

2006-01-19 dwelch
    * Enhanced the status history table in hp-info
    * Put in a temp. fix for the r-value error if using a network connection (r-values not read over network)
    * Added an icon for the Photosmart 8250
    * Reverted to earlier code in status.py that seems to make more LaserJet status work
    * Added a "quick" device query that will be used to make a check to see if the device is not busy or in an error state before various operations 
    
2006-01-18 dwelch
    * Re-fixed defect 750 (The other dialing members of the Groups are not received fax except the first dialing recipient while sending a fax to Groups .)[single file send to group failed]
    * Re-added some icons that were removed by mistake (small error icon and cyan supplies icon) [defect 821(c)]
    * Changed device list in toolbox to use small icons for status overlay

2006-01-17 dwelch
    * Fixed defect 821(b) (Missing down arrow graphic in sendfax)
    * Fixed defect 825 (The "Configure HP Device Manager Setting" dialog can not be launched.)
    * Fixed defect 824 (Coverpage add button not staying disabled)

2006-01-13 dwelch
    * Release 0.9.8.16

2006-01-13 dwelch
    * Reverted fax code back to version HPLIP 0.9.8.14 (removed most changes made on 2006-01-10)
    * Added logfile capability to logger and hp-sendfax
    * Increased error handling in sections of the fax send code
    
2006-01-12 dwelch
    * Fixed defect 821(a) (The corresponding icons for "Configure Print Settings" and "Perform Color Calibration" are missing in toolbox.)

2006-01-11 dwelch
    * Fixed defect 791 (An error occurs while clicking the "Test" button in "Email Alerts" tab of Configure Settings dialog.)
    * Resolved defect 792 as WORKSFORME ((LJ3380) The fax could not be sent while sending the file with coverpage from OpenOffice.)
    
2006-01-11 dwelch
    * Release 0.9.8.15

2006-01-11 syie
    * Add foomatic and PPD for Feb 06 products.

2006-01-10 syie
    * Modified models.xml file to add pen info back to OJ5500 and disable front panel.
    * Remove Fax support from CLJ4730 and LJ4345.

2006-01-10 dwelch
    * Fixed defect 795 (A redundant blank page and an error log report were printed out in recipients when sending a fax job (includes coverpage and two or more files).)
    * Reworked fax send page state machine to better handle state transitions
    * Fixed defect 811 ((DJ6540) The printer has no response when performing a "Align Cartridges" operation from toolbox.

2006-01-09 dsuffield
    * Removed probe.py from distribution.
	
2006-01-06 dwelch
    * Added an icon for PS3210/PS3310/PS8250
    * Added an icon for DJ32xx
    * Added an icon for LJ24xx
    * Added an icon for OJ5xx/6xx/7xx
    * Added an icon for OJ56xx
    * Added an icon for OJK80/K60
    * Added an icon for PS8050
    * Added an icon for PS8750
    * Added an icon for OJK550/K850
    * Removed duplicate device icons and rewroked ones that were not good looking (removed ~160 icons)
    * Readjusted models.xml, etc. to use smaller subset of device icons
    * Fixed the model editor so that it can handle the case where the icon file cannot be found (v6.1)

2006-01-05 dwelch
    * Fixed defect 682 ((PSC5610) Nothing will be done if we select "next" button to do the align cartridge function)
    * Added checks to each cmdline utility and each toolbox function to make sure printer is not busy or in error state before attempting operation
    * Fixed defect 711 (An error occurs while executing the command "hp-check -lnone" in terminal.)
    * Fixed defect 716 (Unwanted message is displayed in the usage of "hp-check" in terminal)

2006-01-05 syie
    * Modified models.xml to remove fax from PSC 1600.

2006-01-04 rcauligi
    * Removed a duplicate ; in ljjetready.h
    * Commented out forward declaration of jpeg APIs from ljjetready.cpp - some compilers don't compile otherwise

2006-01-03 syie
    * Modified models.xml to remove photocard and fax selection for PSC1500.

prnt/hpijs/hpijs.cpp
/*****************************************************************************\
    hpijs.cpp : HP Inkjet Server

    Copyright (c) 2001 - 2004, Hewlett-Packard Co.
    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:
    1. Redistributions of source code must retain the above copyright
       notice, this list of conditions and the following disclaimer.
    2. Redistributions in binary form must reproduce the above copyright
       notice, this list of conditions and the following disclaimer in the
       documentation and/or other materials provided with the distribution.
    3. Neither the name of the Hewlett-Packard nor the names of its
       contributors may be used to endorse or promote products derived
       from this software without specific prior written permission.

    THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
    IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
    OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
    IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
    NOT LIMITED TO, PATENT INFRINGEMENT; PROCUREMENT OF SUBSTITUTE GOODS OR
    SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
    HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
    STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
    IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGE.
\*****************************************************************************/

#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <syslog.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include "header.h"
#include "ijs.h"
#include "ijs_server.h"
#include "services.h"
#include "hpijs.h"

#ifdef HAVE_LIBHPIP
extern  int hpijsFaxServer (int argc, char **argv);
#endif

extern int hpijs97();

int bug(const char *fmt, ...)
{
   char buf[256];
   va_list args;
   int n;

   va_start(args, fmt);

   if ((n = vsnprintf(buf, 256, fmt, args)) == -1)
      buf[255] = 0;     /* output was truncated */

   fprintf(stderr, buf);
   syslog(LOG_WARNING, buf);

   fflush(stderr);
   va_end(args);
   return n;
}

/* Set Print Context. */
int hpijs_set_context(UXServices *pSS)
{
   int r;

   if (pSS->PenSet != DUMMY_PEN)
   {
     if ((r = pSS->pPC->SetPenSet((PEN_TYPE)pSS->PenSet)) != NO_ERROR)
        bug("unable to SetPenSet set=%d, err=%d\n", pSS->PenSet, r);
   }

   if ((r = pSS->pPC->SelectPrintMode((QUALITY_MODE)pSS->Quality, (MEDIATYPE)pSS->MediaType, (COLORMODE)pSS->ColorMode)) !=  NO_ERROR)
   {
      BOOL        bDevText;
      bug("unable to set Quality=%d, ColorMode=%d, MediaType=%d, err=%d\n", pSS->Quality, pSS->ColorMode, pSS->MediaType, r);
      pSS->pPC->GetPrintModeSettings((QUALITY_MODE &)pSS->Quality, (MEDIATYPE &)pSS->MediaType, (COLORMODE &)pSS->ColorMode, bDevText);
      bug("following will be used Quality=%d, ColorMode=%d, MediaType=%d\n", pSS->Quality, pSS->ColorMode, pSS->MediaType);
   }

   /* Map ghostscript paper size to APDK paper size. */
   pSS->MapPaperSize(pSS->PaperWidth, pSS->PaperHeight); 

   /* Do duplex stuff now, since we have a valid print mode. */
   if (pSS->Duplex && !pSS->Tumble)
   {
      if (pSS->pPC->SelectDuplexPrinting(DUPLEXMODE_BOOK) != TRUE)
         bug("unable to set duplex mode=book\n");
   }
   else if (pSS->Duplex && pSS->Tumble)
   {
      if (pSS->pPC->SelectDuplexPrinting(DUPLEXMODE_TABLET) != TRUE)
         bug("unable to set duplex mode=tablet\n");
   }
   else
   {
      pSS->pPC->SelectDuplexPrinting(DUPLEXMODE_NONE);
   }

   if (pSS->MediaPosition != sourceTrayAuto)
   {
      if (pSS->pPC->SetMediaSource((MediaSource)pSS->MediaPosition) != NO_ERROR)
         bug("unable to set MediaPosition=%d\n", pSS->MediaPosition);
   }

   return 0;
}

int hpijs_status_cb(void *status_cb_data, IjsServerCtx *ctx, IjsJobId job_id)
{
   return 0;
}

int hpijs_list_cb(void *list_cb_data, IjsServerCtx *ctx, IjsJobId job_id,
        char *val_buf, int val_size)
{
   return snprintf(val_buf, val_size, "OutputFD,DeviceManufacturer,DeviceModel,PageImageFormat,Dpi,Width,Height,BitsPerSample,ColorSpace,PaperSize,PrintableArea,PrintableTopLeft,PS:Duplex,PS:Tumble,Quality:Quality,Quality:MediaType,Quality:ColorMode,Quality:PenSet,Quality:FullBleed,PS:MediaPosition");
}

int hpijs_enum_cb(void *enum_cb_data, IjsServerCtx *ctx, IjsJobId job_id,
       const char *key, char *val_buf, int val_size)
{
   UXServices *pSS = (UXServices*)enum_cb_data;

   if (!strcmp (key, "ColorSpace"))
   {
      if (pSS->pPC->SupportSeparateBlack())
         return snprintf(val_buf, val_size, "sRGB,KRGB");
      else
         return snprintf(val_buf, val_size, "sRGB");
   }
   else if (!strcmp (key, "DeviceManufacturer"))
      return snprintf(val_buf, val_size, "HEWLETT-PACKARD,APOLLO,HP");
   else if (!strcmp (key, "PageImageFormat"))
      return snprintf(val_buf, val_size, "Raster");
   else if (!strcmp (key, "BitsPerSample"))
      return snprintf(val_buf, val_size, "8");
   else if (!strcmp (key, "PS:Duplex"))
   {
      if (pSS->pPC->SelectDuplexPrinting(DUPLEXMODE_BOOK) == TRUE)
         return snprintf(val_buf, val_size, "true");
      else
         return snprintf(val_buf, val_size, "false");
      pSS->pPC->SelectDuplexPrinting(DUPLEXMODE_NONE);
   }
   else if (!strcmp (key, "PS:Tumble"))
   {
      if (pSS->pPC->SelectDuplexPrinting(DUPLEXMODE_TABLET) == TRUE)
         return snprintf(val_buf, val_size, "true");
      else
         return snprintf(val_buf, val_size, "false");
      pSS->pPC->SelectDuplexPrinting(DUPLEXMODE_NONE);
   }
   else
      bug("unable to enum key=%s\n", key);    
  return IJS_ERANGE;
}

/*
 * Set parameter (in the server) call back. Note, OutputFD is the only call that can be
 * preceded by set DeviceManufacturer and DeviceModel.
 */
int hpijs_set_cb (void *set_cb_data, IjsServerCtx *ctx, IjsJobId job_id,
        const char *key, const char *value, int value_size)
{
   UXServices *pSS = (UXServices*)set_cb_data;
   int fd, r;
   char *tail;
   int status = 0;
   char svalue[IJS_MAX_PARAM+1];   
   float w, h;

   /* Sanity check input value. */
   if (value_size > IJS_MAX_PARAM)
   {
     memcpy(svalue, value, IJS_MAX_PARAM);
     svalue[IJS_MAX_PARAM] = 0;
   }
   else
   {
     memcpy(svalue, value, value_size);
     svalue[value_size] = 0;
   }

   if (!strcmp (key, "OutputFD"))
   {
      fd = strtol(svalue, &tail, 10);
      pSS->OutputPath = fd;   /* set prn_stream as output of SS::ToDevice */
   }
   else if (!strcmp (key, "DeviceManufacturer"))
   {
      if ((strncasecmp(svalue, "HEWLETT-PACKARD", 15) != 0) &&
         (strncasecmp(svalue, "APOLLO", 6) != 0) && (strncasecmp(svalue, "HP", 2) != 0))
      {
         bug("unable to set DeviceManufacturer=%s\n", svalue);
         status = -1;
      }
   }
   else if (!strcmp (key, "DeviceModel"))
   {
      if ((r = pSS->pPC->SelectDevice(svalue)) != NO_ERROR)
      {
         /* OfficeJet LX is not very unique, do separate check here. */
         if (!strncmp(svalue,"OfficeJet", 10))
	    r = pSS->pPC->SelectDevice("DESKJET 540");
      }

      if (r == NO_ERROR)
      {
         pSS->Model = 1;

         /* Got a valid device class, let's set some print mode defaults. */
         BOOL        bDevText;
         pSS->pPC->GetPrintModeSettings((QUALITY_MODE &)pSS->Quality, (MEDIATYPE &)pSS->MediaType, (COLORMODE &)pSS->ColorMode, bDevText);
      }
      else
      {
         bug("unable to set device=%s, err=%d\n", svalue, r);
         status = -1;
      }
   }
   else if ((strcmp (key, "PS:Duplex") == 0) || (strcmp (key, "Duplex") == 0))
   {
      if (strncmp(svalue, "true", 4) == 0)
         pSS->Duplex = 1;
      else
         pSS->Duplex = 0;
   }
   else if ((strcmp (key, "PS:Tumble") == 0) || (strcmp (key, "Tumble") == 0))
   {
      if (strncmp(svalue, "true", 4) == 0)
         pSS->Tumble = 1;
      else
         pSS->Tumble = 0;
   }
   else if (!strcmp (key, "PaperSize"))
   {
      w = (float)strtod(svalue, &tail);
      h = (float)strtod(tail+1, &tail);

      if (pSS->FirstRaster)
      {
         /* Normal start of print Job. */
         pSS->PaperWidth = w;
         pSS->PaperHeight = h;
         hpijs_set_context(pSS);
      }
      else
      {
         /* Middle of print Job, ignore paper size if same. */
         if (!(w == pSS->PaperWidth && h == pSS->PaperHeight))
         {
            pSS->FirstRaster = 1;  /* force new Job */
            pSS->PaperWidth = w;   /* set new paper size */
            pSS->PaperHeight = h;
            hpijs_set_context(pSS);
         }
      }
   }
   else if (!strcmp (key, "TopLeft"))
   {
      /* not currently used */
   }
   else if (!strcmp (key, "Quality:Quality"))
   {
      pSS->Quality = strtol(svalue, &tail, 10);
   }
   else if (!strcmp (key, "Quality:MediaType"))
   {
      pSS->MediaType = strtol(svalue, &tail, 10);
   }
   else if (!strcmp (key, "Quality:ColorMode"))
   {
      pSS->ColorMode = strtol(svalue, &tail, 10);
   }
   else if (!strcmp (key, "Quality:PenSet"))
   {
      pSS->PenSet = strtol(svalue, &tail, 10);
   }
   else if (!strcmp (key, "Quality:FullBleed"))
   {
      pSS->FullBleed = strtol(svalue, &tail, 10);
   }
   else if (!strcmp (key, "PS:MediaPosition"))
   {
      pSS->MediaPosition = strtol(svalue, &tail, 10);
   }
   else
      bug("unable to set key=%s, value=%s\n", key, svalue);    

   return status;
}

/* Get parameter (from the server) call back. Note, all calls must be preceded by set DeviceName. */
int hpijs_get_cb(void *get_cb_data, IjsServerCtx *ctx, IjsJobId job_id, const char *key, char *value_buf, int value_size)
{
   UXServices *pSS = (UXServices*)get_cb_data;
   float        fY;

   if (!strcmp (key, "PrintableArea"))
   {
       fY = pSS->pPC->PrintableHeight ();
      /* If duplexing, adjust printable height to 1/2 inch top/bottom margins, except laserjets. */
      if ((pSS->pPC->QueryDuplexMode() != DUPLEXMODE_NONE) && pSS->pPC->RotateImageForBackPage())
      {
          // User has requested fullbleed printing and printer supports 4-sided fullbleed or
          // top and bottom margin are equal (0.125"), then
          // don't need top and bottom margins to be 0.5" each. In other words, if physical page size -
          // printable height is more than 0.25 inches, adjustment is required for symmetry.
          // Update - 11/22/05: Not so, the autoduplexer still requires a half inch margin.
#if 0
         if ((pSS->pPC->PrintableHeight () + 0.28) < pSS->pPC->PhysicalPageSizeY ())
         {
             fY = pSS->pPC->PhysicalPageSizeY () - 1.0;
         }
#endif
         fY = pSS->pPC->PhysicalPageSizeY () - 1.0;
//         return snprintf(value_buf, value_size, "%.4fx%.4f", pSS->pPC->PrintableWidth(), pSS->pPC->PhysicalPageSizeY()-1);
      }
//      else
//         return snprintf(value_buf, value_size, "%.4fx%.4f", pSS->pPC->PrintableWidth(), pSS->pPC->PrintableHeight());
      return (snprintf (value_buf, value_size, "%.4fx%.4f", pSS->pPC->PrintableWidth (), fY));
   }
   else if (!strcmp (key, "PrintableTopLeft"))
   {
       fY = pSS->pPC->PrintableStartY (); 
      /* If duplexing, adjust printable top to 1/2 inch top margin, except laserjets. */
      if ((pSS->pPC->QueryDuplexMode() != DUPLEXMODE_NONE) && pSS->pPC->RotateImageForBackPage())
      {
#if 0
         if ((pSS->pPC->PrintableHeight () + 0.28) < pSS->pPC->PhysicalPageSizeY ())
         {
             fY = 0.5;
         }
#endif
         fY = 0.5;
      }
      return snprintf (value_buf, value_size, "%.4fx%.4f", pSS->pPC->PrintableStartX (), fY);

/*
         return snprintf(value_buf, value_size, "%.4fx%.4f", pSS->pPC->PrintableStartX(), 0.5);
      else
         return snprintf(value_buf, value_size, "%.4fx%.4f", pSS->pPC->PrintableStartX(), pSS->pPC->PrintableStartY());
 */
   }
   else if ((!strcmp (key, "Duplex")) || (!strcmp (key, "PS:Duplex")))
   {
      if (pSS->pPC->QueryDuplexMode() == DUPLEXMODE_NONE)
         return snprintf(value_buf, value_size, "false");
      else
         return snprintf(value_buf, value_size, "true");
   }
   else if ((!strcmp (key, "Tumble")) || (!strcmp (key, "PS:Tumble")))
   {
      if (pSS->pPC->QueryDuplexMode() == DUPLEXMODE_TABLET)
         return snprintf(value_buf, value_size, "true");
      else
         return snprintf(value_buf, value_size, "false");
   }
   else if (!strcmp (key, "PaperSize"))
   {
      return snprintf(value_buf, value_size, "%.4fx%.4f", pSS->pPC->PhysicalPageSizeX(), pSS->pPC->PhysicalPageSizeY());
   }
   else if (!strcmp (key, "Dpi"))
   {
      return snprintf(value_buf, value_size, "%dx%d", pSS->pPC->EffectiveResolutionX(), pSS->pPC->EffectiveResolutionY());
   }
   else if (!strcmp (key, "DeviceModel"))
   {
      return snprintf(value_buf, value_size, "%s", pSS->pPC->PrinterModel());
   }
   else if (!strcmp (key, "Quality:Quality"))
   {
      return snprintf(value_buf, value_size, "%d", pSS->Quality);
   }
   else if (!strcmp (key, "Quality:ColorMode"))
   {
      return snprintf(value_buf, value_size, "%d", pSS->ColorMode);
   }
   else if (!strcmp (key, "Quality:MediaType"))
   {
      return snprintf(value_buf, value_size, "%d", pSS->MediaType);
   }
   else if (!strcmp (key, "ColorSpace"))
   {
      return snprintf(value_buf, value_size, pSS->ph.cs);
   }
   else if (!strcmp (key, "PageImageFormat"))
   {
      return snprintf(value_buf, value_size, "Raster");
   }
   else if (!strcmp (key, "BitsPerSample"))
   {
      return snprintf(value_buf, value_size, "8");
   }
   else if (!strcmp (key, "PS:MediaPosition"))
   {
      return snprintf(value_buf, value_size, "%d", pSS->MediaPosition);
   }
   else
      bug("unable to get key=%s\n", key);    

   return IJS_EUNKPARAM;
}

/* Get raster from the client. */
int hpijs_get_client_raster(IjsServerCtx *ctx, char *buf, int size, char white)
{
   int status, clean=1, i;

   status = ijs_server_get_data(ctx, (char *)buf, size);

   if (status < 0)
      return status;  /* error */

   /* Check for blank raster. */
   for (i = 0; i < size; i++)
   {
      if (buf[i] != white)
      {
         clean = 0;
         break;
      }
   }

   if (clean)
      return 0;

   return size;
}

int main(int argc, char *argv[], char *evenp[])
{
   UXServices *pSS = NULL;
   IjsServerCtx *ctx = NULL;
   char *raster = NULL, *k_raster = NULL;
   int status = EXIT_FAILURE;
   int ret, n, i, kn=0, width, k_width;

   if (argc > 1)
   {
      const char *arg = argv[1];
      if ((arg[0] == '-') && (arg[1] == 'h'))
      {
         fprintf(stdout, "\nHewlett-Packard Co. Inkjet Server %s\n", VERSION);
         fprintf(stdout, "Copyright (c) 2001-2004, Hewlett-Packard Co.\n");
         exit(0);
      }
   }

#ifdef HAVE_LIBHPIP
   char *pDev;
   if ((pDev = getenv ("DEVICE_URI")) &&
       ((strncmp (pDev, "hpfax:", 6)) == 0))
   {
       exit ( hpijsFaxServer (argc, argv));
   }
#endif

   ctx = ijs_server_init();
   if (ctx == NULL)
   {
      bug("unable to init hpijs server\n");
      goto BUGOUT;
   }

   pSS = new UXServices();
   if (pSS->constructor_error != NO_ERROR)
   {
      bug("unable to open Services object err=%d\n", pSS->constructor_error);
      goto BUGOUT;
   }

#ifdef CAPTURE
   if ((pSS->InitScript("/tmp/capout", TRUE)) != NO_ERROR)
      bug("unable to init capture");
#endif

   pSS->pPC = new PrintContext(pSS, 0, 0);

   /* Ignore JOB_CANCELED. This a bi-di hack that allows the job to continue even if bi-di communication failed. */
   if (pSS->pPC->constructor_error > 0 && pSS->DisplayStatus != DISPLAY_PRINTING_CANCELED)
   {
      bug("unable to open PrintContext object err=%d\n", pSS->pPC->constructor_error);
      goto BUGOUT;
   }
   if (pSS->pPC->constructor_error < 0)
      bug("WARNING: %s\n", pSS->GetDriverMessage(pSS->pPC->constructor_error));

#if 0
   bug("device model=%s\n", pSS->pPC->PrinterModel());
   bug("device class=%s\n",  pSS->pPC->PrintertypeToString(pSS->pPC->SelectedDevice()));
   bug("default pen=%d\n",  pSS->pPC->GetDefaultPenSet());
   bug("installed pen=%d\n",  pSS->pPC->GetInstalledPens());
#endif

   ijs_server_install_status_cb (ctx, hpijs_status_cb, pSS);
   ijs_server_install_list_cb (ctx, hpijs_list_cb, pSS);
   ijs_server_install_enum_cb (ctx, hpijs_enum_cb, pSS);
   ijs_server_install_set_cb (ctx, hpijs_set_cb, pSS);
   ijs_server_install_get_cb (ctx, hpijs_get_cb, pSS);

   while (1)
   {
      if ((ret = ijs_server_get_page_header(ctx, &pSS->ph)) < 0)
      {
         bug("unable to read client data err=%d\n", ret);
         goto BUGOUT;
      }

      if (pSS->Model == -1)
         goto BUGOUT;      /* no device selected */

      if (ret)
      {
         status = 0; /* normal exit */
         break;
      }

      if (pSS->FirstRaster)
      {
         pSS->FirstRaster = 0;

         width = (int)(pSS->ph.xres * pSS->pPC->PrintableWidth() + 0.5);

         /* Desensitize input width, may be off by one due to paper size conversions. */
         if (pSS->ph.width < width)
            width = pSS->ph.width;

         if ((ret = pSS->pPC->SetPixelsPerRow(width, width)) != NO_ERROR)
         {
            bug("unable to SetPixelsPerRow width=%d, err=%d\n", pSS->ph.width, ret);
         }

         /* Turn off any bi-di support. Allow bi-di for printer capabilities only. */
//         pSS->IOMode.bDevID = pSS->IOMode.bStatus = FALSE;
         pSS->ResetIOMode (FALSE, FALSE);

         if (pSS->pJob != NULL)
            delete pSS->pJob;
         pSS->pJob = new Job(pSS->pPC);
         if (pSS->pJob->constructor_error != NO_ERROR)
         {
            bug("unable to create Job object err=%d\n", pSS->pJob->constructor_error);
            goto BUGOUT;
         }

         if (pSS->pPC->QueryDuplexMode() != DUPLEXMODE_NONE)
         {
            if ((pSS->pPC->QueryDuplexMode() == DUPLEXMODE_BOOK) && pSS->pPC->RotateImageForBackPage())
               pSS->InitDuplexBuffer();
            pSS->BackPage = FALSE;
         }

         pSS->KRGB=0;
         if (strcmp(pSS->ph.cs, "KRGB") == 0)
            pSS->KRGB=1;
         
#if 0
         bug("papersize=%d\n", pSS->pPC->GetPaperSize());
         bug("width=%d, height=%d\n", pSS->ph.width, pSS->ph.height);
         bug("EffResX=%d, EffResY=%d, InPixelsPerRow=%d, OutPixelsPerRow=%d\n", 
            pSS->pPC->EffectiveResolutionX(), pSS->pPC->EffectiveResolutionY(), 
            pSS->pPC->InputPixelsPerRow(), pSS->pPC->OutputPixelsPerRow());
         bug("device=%s\n", pSS->pPC->PrinterModel());
#endif
      } // pSS->FirstRaster

      if ((raster = (char *)malloc(pSS->ph.width*3)) == NULL)
      {
         bug("unable to allocate raster buffer size=%d: %m\n", pSS->ph.width*3);
         goto BUGOUT;
      }

      k_width = (pSS->ph.width + 7) >> 3;  /* width of k plane in bytes, byte aligned */

      if ((k_raster = (char *)malloc(k_width)) == NULL)
      {
         bug("unable to allocate black raster buffer size=%d: %m\n", k_width);
         goto BUGOUT;
      }
      memset(k_raster, 0, k_width);

      for (i=0; i < pSS->ph.height; i++)      
      {
         if ((n = hpijs_get_client_raster(ctx, raster, pSS->ph.width*3, 0xff)) < 0)
            break;    /* error */
         if (pSS->KRGB)
         {
            if ((kn = hpijs_get_client_raster(ctx, k_raster, k_width, 0)) < 0)
               break;    /* error */
         }
         if (n == 0 && kn == 0)
            pSS->ProcessRaster((char *)0, (char *)0);  /* blank raster */
         else if (kn == 0)
            pSS->ProcessRaster(raster, (char *)0);
         else if (n == 0)
            pSS->ProcessRaster((char *)0, k_raster);
         else 
            pSS->ProcessRaster(raster, k_raster);
      }

      free(raster);
      raster = NULL;
      free(k_raster);
      k_raster = NULL;

      if (pSS->pPC->QueryDuplexMode() != DUPLEXMODE_NONE)
      {
         if ((pSS->pPC->QueryDuplexMode() == DUPLEXMODE_BOOK) && pSS->pPC->RotateImageForBackPage() && pSS->BackPage)
         {
            pSS->SendBackPage();
         }
         pSS->BackPage = (BOOL)((int)pSS->BackPage + 1) % 2;
      }

      pSS->pJob->NewPage();

   } /* end while (1) */

   if (pSS->pPC->QueryDuplexMode() != DUPLEXMODE_NONE)
   {
      if (pSS->BackPage)
      {
         /* Send extra blank line & newpage to eject the page. (for VIP printers). */
         /* For malibu send enough blank lines to cause at least two blank rasters in Job::sendrasters(). 5/1/03, des */
         //         for (int i = 0; i < 201; i++)
         for (int i = 0; i < 401; i++)
            pSS->pJob->SendRasters((unsigned char *)0);
         pSS->pJob->NewPage();
      }
   }

BUGOUT:
   if (pSS != NULL)
   {
      if (pSS->pJob != NULL)
         delete pSS->pJob;
      if (pSS->pPC != NULL)
         delete pSS->pPC;
#ifdef CAPTURE
      pSS->EndScript();
#endif
      delete pSS;
   }
   if (raster != NULL)
      free(raster);
   if (k_raster != NULL)
      free(k_raster);
   if (ctx != NULL)
      ijs_server_done(ctx);

   exit(status);
}

